### ISSUE TITLE
changing a field value based on state of another one (for repeated, nesting field)

### ISSUE THREAD

**Issue by faassen** [foxhound87/mobx-ajv-form#200]
In some cases I want to change the field value (in particular clear it) based on the state of another value.

I think the right way in would be ```form.on('update@state.city',..``` and then use a ```clear()``` on the field I want to clear. But my form is a nested form, and I tried ```form.on('update@repeated[].value', ...``` with no luck: it never seems to fire.

As an alternative I tried to hook up events using ```form.$('repeated[].value').onChange```, but the field couldn't be found in that case.

Two questions:

* are these strategies supposed to work, or am I going about it the wrong way?

* is this a bug concerning nested fields and events?

**Comment by faassen** [comment]
The ``onChange`` won't work because I think these are the default event handlers that mobx-react-form uses to hook up to an input element. But unfortunately ``on`` doesn't seem to work for nested fields.

**Comment by faassen** [comment]
I think the event system generates a path for the field, but that will always have an index in it in the case of repeated values. There therefore doesn't appear to be a way to listen to repeated values. So I think the event system should normalize paths to strip out indices to make this work.

**Comment by foxhound87** [comment]
You should look at this open issue: https://github.com/foxhound87/mobx-react-form/issues/172

If you scroll, there's an example about using mobx **observe** method on the fields, used in conjunction of the package's `forEach()` method that will deeply iterate all fields and nested fields.

I would like to add that as all the fields are **mobx observables maps** they have also special methods as **intercept** and **observe**, you can find more [on the mobx docs](https://mobx.js.org/refguide/map.html). 

I still searching a good way to re-implement the package's Event Hooks using these methods, but in the meanwhile you can figure out how to use them in your project.

**Comment by foxhound87** [comment]
You can access the **obsevable map** through the `fields` props in both **Form** and **Field** instances: `form.fields` / `field.fields`.

**Comment by foxhound87** [comment]
Or if you need to reimplement the [Event Handlers](https://foxhound87.github.io/mobx-react-form/docs/events/events-handlers.html) you can follow my last blog article:

[Validate Field onBlur() using Props Bindings in mobx-react-form](https://medium.com/@foxhound87/validate-field-onblur-using-props-bindings-in-mobx-react-form-3de48ba00576)

**Comment by faassen** [comment]
Thanks, I'll take a look at mobx itself. I should have thought of that. It's another use case to cover in narrative documentation.

**Comment by faassen** [comment]
I took a look at using observable, but there is some complexity here in combination with repeated sub-forms.

I can implement an ``onInit`` that uses ``forEach`` to go through all fields, including nested ones. These fields become observable. It's not the most expressive way to make such fields observable (I'd prefer to use the flat specification, like ``bars[].value``), but it could be made to work.

But what happens now if someone adds a field using ``onAdd``? This field is not observable. Instead I have to write more logic to make the values that just got added observable. I'm also not sure what would happen if I remove a field previously declared observable -- maybe there's no problem there.

I see you closed this issue, and I hope I missed something, but this isn't a very pleasant API when dealing with repeating fields.

I think it would be useful to have a declarative callback, something like:

```
observers = {
  'entries[].a': (field, parentField, form) =>  { parentField.$('b').value = ""; }
}
```

To express the following relationship: blank the value ``b`` on the entry with the same index when ``entries[].a`` changes.

I will now investigate the other approach you suggested, where I override the event handler.

**Comment by faassen** [comment]
See also #202 where I discuss a way to navigate to the parent field, which would make the signature of the calback simpler.

**Comment by foxhound87** [comment]
The idea of custom observers for each field is very cool.

**Comment by foxhound87** [comment]
I'm working on a different kind of implementations for observers, 

something like: 

"Use observe() on the selected field, specify the prop key to observe, and the function to execute."

```javascript
    this.$('members').observe({
      key: 'value',
      call: data => console.log('changed', data),
    });
```

With this method you can observe not only values but also other properties, like `focus`, `touched`, `changed` and so on...

And If you specify the `key` as `fields`, it will observer the mobx observable map (`add` and `delete` events) of the nested fields instead the field props.

The `data` in in the callback will have the form instance, the selected field instance, the field path, and the mobx object of the event state (oldValue, newValue, type of the mobx event and other stuff based on the mobx observable type).

Eventually, if you call observe() from the Form instance, you can pass an additional `path` for selecting the field.

Sounds good?

**Comment by foxhound87** [comment]
Implemented in [version 1.19](https://github.com/foxhound87/mobx-react-form/commit/005d04c7)

[Observe Fields Props](https://foxhound87.github.io/mobx-react-form/docs/events/observe.html)

**Comment by foxhound87** [comment]
Anyway both styles can be used.
You can use the 'observables' initial object as well, this will allow to enable the observables to the new added fields.
